### PR TITLE
Superquadric tensor fix

### PR DESCRIPTION
**Pull request by RubioJr9**
- Superquadric glyphs need to change axis alignment if they're more linear than planar.
Before the fix:
superquadric_bug
After the fix:
superquadric_fixed

- Changed superquadric emphasis range from 0-1 to 0-10. We were limitied to very ellipsoidal shapes.
- Changed superquadric emphasis default from 0.85 to 3. The creator used 3 and 6 as examples and said [3 is a a good default choice](http://www.cs.utah.edu/~gk/papers/vissym04/web/slide020.html). 

- I incorrectly named superquadric tensors as superellipsoids so I fixed it. [Superellipsoids](https://en.wikipedia.org/wiki/Superellipsoid) are a subset of [superquadric surfaces](https://en.wikipedia.org/wiki/Superquadrics). I thought superellipsoid was a more specific name, however the creator named it [superquadric tensor glyphs](http://www.cs.utah.edu/~gk/papers/vissym04/web/index.html). I changed the name in ShowFieldGlyphs but left in a line so old files will still display the same.

Edit: The above links to the slides are dead. The paper can be found here http://www.sci.utah.edu/publications/gk04/vissym04kindlmann.pdf

### CODE REVIEW

**Comment by dcwhite** [comment]
@RubioJr9 CodeFactor flagged the new method as too complex. I'm going to agree with the failure on this one--it should be refactored to be smaller. There are a lot of duplicated chunks that can be pulled out.

https://github.com/SCIInstitute/SCIRun/blob/92a424a4aa5e73ff7bb98859b64cba41515f3722/src/Graphics/Glyphs/GlyphGeom.cc#L996-L1137

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 1]
```diff
@@ -1057,8 +1060,9 @@ void GlyphGeom::generateSuperEllipsoid(const Point& center, Tensor& t, double sc
 
   double cl = (eigvals[0] - eigvals[1]) / (eigvals[0] + eigvals[1] + eigvals[2]);
   double cp = 2.0 * (eigvals[1] - eigvals[2]) / (eigvals[0] + eigvals[1] + eigvals[2]);
-  double A = spow((1.0 - cl), emphasis);
-  double B = spow((1.0 - cp), emphasis);
+  bool linear = cl >= cp;
+  double A = linear ? spow((1.0 - cp), emphasis) : spow((1.0 - cl), emphasis);
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]: Some more specific critiques.

Don't do computation inside a ternary operator. They should be used sparingly, in simple expression. For this case, pull out the part that actually varies--the cp/cl variable--and put that in a ternary if you must, then do the computation once in code. Above all, try to ascribe to DRY--Don't Repeat Yourself.

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 2]
```diff
@@ -1077,27 +1081,47 @@ void GlyphGeom::generateSuperEllipsoid(const Point& center, Tensor& t, double sc
...
-        const float nnx = spow(nr[i], 2.0-B) * spow(nx, 2.0-A);
-        const float nny = spow(nr[i], 2.0-B) * spow(ny, 2.0-A);
-        const float nnz = spow(nz[i], 2.0-B);
-        Vector normal = rotate * Vector( nnx, nny, nnz );
+        double nnx, nny, nnz;
+        if(linear)
+        {
+          nnx =  spow(nz[i], 2.0-B);
+          nny = -spow(nr[i], 2.0-B) * spow(ny, 2.0-A);
+          nnz =  spow(nr[i], 2.0-B) * spow(nx, 2.0-A);
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]: Similar pattern of computation as lines 1090-1098.

**Comment by RubioJr9** [comment]
> @RubioJr9 CodeFactor flagged the new method as too complex. I'm going to agree with the failure on this one--it should be refactored to be smaller. There are a lot of duplicated chunks that can be pulled out.
> 
> https://github.com/SCIInstitute/SCIRun/blob/92a424a4aa5e73ff7bb98859b64cba41515f3722/src/Graphics/Glyphs/GlyphGeom.cc#L996-L1137

I just pushed a commit. Let me know what you think of the TensorGlyphBuilder class. I reduced the redundancy quite a bit. I'm going to need to be able to evaluate specific points of the superquadric surface so the refactor of that function was going to be needed.

I would also like to do the same and make builders for vectors and scalars. I think the GeomData could be expanded upon to take the function buildObject() from GlyphGeom. That would make GlyphGeom only have addGlyph() functions so that could be renamed to GlyphFactory.

**Comment by dcwhite** [comment]
Neat, CodeFactor is happy. I will look more closely tomorrow!

**Comment by dcwhite** [comment]
Another Windows build error: https://github.com/SCIInstitute/SCIRun/runs/489264304?check_suite_focus=true#step:6:4413

File: src/Core/GeometryPrimitives/Tensor.cc [hunk 3]
```diff
@@ -380,6 +387,26 @@ void Tensor::set_outside_eigens(const Vector &e1, const Vector &e2,
   have_eigens_ = 1;
 }
 
+double Tensor::eigenValueSum()
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Tensor.cc]: These can all be `const`

File: src/Core/GeometryPrimitives/Tensor.cc [hunk 4]
```diff
@@ -380,6 +387,26 @@ void Tensor::set_outside_eigens(const Vector &e1, const Vector &e2,
...
+{
+  return l1_ + l2_ + l3_;
+}
+
+double Tensor::linearCertainty()
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Tensor.cc]:
```suggestion
double Tensor::linearCertainty() const
```

File: src/Core/GeometryPrimitives/Tensor.cc [hunk 5]
```diff
@@ -380,6 +387,26 @@ void Tensor::set_outside_eigens(const Vector &e1, const Vector &e2,
...
+{
+  return (l1_ - l2_) / eigenValueSum();
+}
+
+double Tensor::planarCertainty()
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Tensor.cc]:
```suggestion
double Tensor::planarCertainty() const
```

File: src/Core/GeometryPrimitives/Tensor.cc [hunk 6]
```diff
@@ -380,6 +387,26 @@ void Tensor::set_outside_eigens(const Vector &e1, const Vector &e2,
...
+{
+  return 2.0 * (l2_ - l3_) / eigenValueSum();
+}
+
+double Tensor::sphericalCertainty()
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Tensor.cc]:
```suggestion
double Tensor::sphericalCertainty() const
```

File: src/Core/GeometryPrimitives/Tensor.h [hunk 7]
```diff
@@ -125,7 +127,12 @@ class SCISHARE Tensor {
   double val(size_t i, size_t j) const { return mat_[i][j]; }
   double& val(size_t i, size_t j) { return mat_[i][j]; }
 
+  double linearCertainty();
```
Comment (dcwhite) [inline src/Core/GeometryPrimitives/Tensor.h]: Add consts here

File: src/Graphics/Glyphs/GeomData.h [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+#define Graphics_Glyphs_GEOM_DATA_H
+
+#include <Core/GeometryPrimitives/Vector.h>
+#include <Core/Datatypes/Color.h>
+#include <Core/Math/TrigTable.h>
+#include <Graphics/Glyphs/share.h>
+
+namespace SCIRun {
+namespace Graphics {
+struct GeomData
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GeomData.h]: Already too many vague names with Geometry and Data...how about `GlyphConstructionParameters`?

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 9]
```diff
@@ -311,29 +316,29 @@ void GlyphGeom::generateCylinder(const Point& p1, const Point& p2, double radius
...
+    geomData_.points.push_back(r1 * p + Vector(p1));
+    geomData_.colors.push_back(color1);
+    geomData_.normals.push_back(normals);
+    geomData_.numVBOElements++;
+    geomData_.points.push_back(r2 * p + Vector(p2));
+    geomData_.colors.push_back(color2);
+    geomData_.normals.push_back(normals);
+    geomData_.numVBOElements++;
+
+    geomData_.indices.push_back(0 + offset);
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]: Now that there's a new class, this block (334-339) can be a function of `GlyphConstructionParameters`. Will again reduce duplication and you can introduce a name to explain the magic numbers.

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 10]
```diff
@@ -311,29 +316,29 @@ void GlyphGeom::generateCylinder(const Point& p1, const Point& p2, double radius
...
-    normals_.push_back(normals);
-    numVBOElements_++;
-
-    indices_.push_back(0 + offset);
-    indices_.push_back(1 + offset);
-    indices_.push_back(2 + offset);
-    indices_.push_back(2 + offset);
-    indices_.push_back(1 + offset);
-    indices_.push_back(3 + offset);
+    geomData_.points.push_back(r1 * p + Vector(p1));
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]: These four lines also make a method--it's the same 4 modifications.

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 11]
```diff
@@ -454,52 +459,52 @@ void GlyphGeom::generateCylinder(const Point& p1, const Point& p2, double radius
       std::sin(strip_angle * strips) * crx;
     p.normalize();
     Vector normals((length * p + (radius2-radius1)*n).normal());
-    points_.push_back(radius1 * p + Vector(p1));
-    colors_.push_back(color1);
-    normals_.push_back(normals);
-    points_.push_back(radius2 * p + Vector(p2));
-    colors_.push_back(color2);
-    normals_.push_back(normals);
+    geomData_.points.push_back(radius1 * p + Vector(p1));
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]: Yeah, these are now method calls on the new class.

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 12]
```diff
@@ -540,14 +545,14 @@ void GlyphGeom::generateSphere(const Point& center, double radius, int resolutio
         v2 = 1;
       }
 
-      indices_.push_back(0 + offset);
-      indices_.push_back(v1 + offset);
-      indices_.push_back(v2 + offset);
-      indices_.push_back(v2 + offset);
-      indices_.push_back(v1 + offset);
-      indices_.push_back(3 + offset);
+      geomData_.indices.push_back(0 + offset);
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]: Another method

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 13]
```diff
@@ -1152,81 +793,81 @@ void GlyphGeom::generateTorus(const Point& p1, const Point& p2, double major_rad
...
 {
-  points_.push_back(Vector(p1));
-  points_.push_back(Vector(p2));
-  points_.push_back(Vector(p3));
-  points_.push_back(Vector(p4));
-  colors_.push_back(color);
-  colors_.push_back(color);
-  colors_.push_back(color);
-  colors_.push_back(color);
+  geomData_.points.push_back(Vector(p1));
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]:
```suggestion
  geomData_.addAllPoints({p1, p2, p3, p4});
```

File: src/Graphics/Glyphs/GlyphGeom.cc [hunk 14]
```diff
@@ -1152,81 +793,81 @@ void GlyphGeom::generateTorus(const Point& p1, const Point& p2, double major_rad
...
+  geomData_.points.push_back(Vector(p2));
+  geomData_.points.push_back(Vector(p3));
+  geomData_.points.push_back(Vector(p4));
+  geomData_.colors.push_back(color);
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeom.cc]:
```suggestion
  geomData_.addColorNTimes(color, 4);
```

File: src/Graphics/Glyphs/GeomData.h (hunk 8 above)
Comment (RubioJr9) [reply src/Graphics/Glyphs/GeomData.h]: I went with GlyphConstructor and moved buildObject() here

File: src/Graphics/Glyphs/GlyphGeom.cc (hunk 9 above)
Comment (RubioJr9) [reply src/Graphics/Glyphs/GlyphGeom.cc]: There's now a function addVertex(point, normal, color) which replaces adding each of those to a vector and also increments numVBOElements

File: src/Graphics/Glyphs/GlyphGeom.cc (hunk 10 above)
Comment (RubioJr9) [reply src/Graphics/Glyphs/GlyphGeom.cc]: There's now a function addIndices() which takes 3 index values to form a triangle. Also that offset value can be set and is held internally now.

**Comment by dcwhite** [comment]
@RubioJr9 Merge conflict with this one.

File: src/Core/GeometryPrimitives/Tensor.h (hunk 7 above)
Comment (dcwhite) [reply src/Core/GeometryPrimitives/Tensor.h]: Still no consts?

File: src/Graphics/Glyphs/GlyphGeomUtility.h [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+inline double GlyphGeomUtility::spow(double e, double x)
+{
+  // This for round off of very small numbers.
+  if (abs(e) < 1.0e-6)
+    e = 0.0;
+
+  if (e < 0.0)
+    return (double)(pow(abs(e), x) * -1.0);
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeomUtility.h]:
```suggestion
    return static_cast<double>(std::pow(std::abs(e), x) * -1.0);
```

File: src/Graphics/Glyphs/GlyphGeomUtility.h [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+  else
+    return (double)(pow(e, x));
```
Comment (dcwhite) [inline src/Graphics/Glyphs/GlyphGeomUtility.h]:
```suggestion
    return std::pow(e, x);
```
No cast needed here?

File: src/Core/GeometryPrimitives/Tensor.h (hunk 7 above)
Comment (RubioJr9) [reply src/Core/GeometryPrimitives/Tensor.h]: I swear I added this. I'll do it now

File: src/Graphics/Glyphs/GlyphGeomUtility.h (hunk 16 above)
Comment (RubioJr9) [reply src/Graphics/Glyphs/GlyphGeomUtility.h]: This function was directly copied over from GlyphGeom. I'll add it since it seems like it should be there

**Comment by dcwhite** [comment]
@jessdtate @allywarner Next

**Comment by dcwhite** [comment]
@jessdtate should still review this one, given his tensor knowledge

**Pull request merged**